vnext/ReactUWP/Views/DynamicAutomationPeer.cpp

```diff
@@ -0,0 +1,190 @@
+    return winrt::AutomationControlType::Image;
+  case winrt::react::uwp::AccessibilityRoles::KeyboardKey:
+    return winrt::AutomationControlType::Custom;
+  case winrt::react::uwp::AccessibilityRoles::Text:
+  case winrt::react::uwp::AccessibilityRoles::Summary:
+  case winrt::react::uwp::AccessibilityRoles::Header:
+    return winrt::AutomationControlType::Text;
+  case winrt::react::uwp::AccessibilityRoles::Adjustable:
+    return winrt::AutomationControlType::Slider;
+  case winrt::react::uwp::AccessibilityRoles::Search:
+  case winrt::react::uwp::AccessibilityRoles::Unknown:
+  default:
+    return winrt::AutomationControlType::Group;
+  }
+
+  return Super::GetAutomationControlTypeCore();
+}
+
+winrt::IInspectable DynamicAutomationPeer::GetPatternCore(winrt::PatternInterface const& patternInterface) const
+{
+  auto accessibilityRole = GetAccessibilityRole();
+
+  if (patternInterface == winrt::PatternInterface::Invoke &&
+    (accessibilityRole == winrt::react::uwp::AccessibilityRoles::Button || accessibilityRole == winrt::react::uwp::AccessibilityRoles::ImageButton))
+  {
+    return *this;
+  }
+  else if (patternInterface == winrt::PatternInterface::Selection || patternInterface == winrt::PatternInterface::SelectionItem)
+  {
+    return *this;
+  }
+
+  return Super::GetPatternCore(patternInterface);
+}
+
+bool DynamicAutomationPeer::IsEnabledCore() const
+{
+  bool disabled = GetAccessibilityState(winrt::react::uwp::AccessibilityStates::Disabled);
+  return !disabled && Super::IsEnabledCore();
+}
+
+// IInvokeProvider
+
+void DynamicAutomationPeer::Invoke() const
+{
+  if (auto invokeHandler = GetAccessibilityInvokeEventHandler())
+  {
+    invokeHandler();
+  }
+}
+
+// ISelectionProvider
+
+winrt::com_array<winrt::IRawElementProviderSimple> DynamicAutomationPeer::GetSelection() const
+{
+  // We don't differentiate between Views that are containers and Views that are items, and we don't
+  // have a way to specify the linkage between the two
+
+  // Returning nothing until that linkage exists
+  return {};
+}
+
+// ISelectionItemProvider
+
+bool DynamicAutomationPeer::IsSelected() const
+{
+  return GetAccessibilityState(winrt::react::uwp::AccessibilityStates::Selected);
+}
+
+winrt::IRawElementProviderSimple DynamicAutomationPeer::SelectionContainer() const
+{
+  // We don't differentiate between Views that are containers and Views that are items, and we don't
+  // have a way to specify the linkage between the two
+
+  // Returning nothing until that linkage exists
+
+  return nullptr;
+}
+
+void DynamicAutomationPeer::AddToSelection() const
+{
+  
+}
+
+void DynamicAutomationPeer::RemoveFromSelection() const
+{
+
+}
+
+void DynamicAutomationPeer::Select() const
+{
+
+}
+
+winrt::react::uwp::AccessibilityRoles DynamicAutomationPeer::GetAccessibilityRole() const
+{
+  try
+  {
+    if (auto owner = Owner())
+    {
+      return DynamicAutomationProperties::GetAccessibilityRole(owner);
+    }
+  }
+  catch (...) {}
```

Comment:
What sorts of exceptions show up here?

Reply:
Same question for GetAccessibilityState and GetAccessibilityInvokeEventHandler

Reply:
We're not reference tracking properly. So when UIA grabs the peer, there's a case that the Owner will have been released. We're using an old cppwinrt - I think it'll be fixed when we use a newer version, or we may need to take the reference tracking stuff that mux uses to implement the correct lifetime behavior. Neither are small tasks.

Reply:
In the meantime, this prevents UIA tools from crashing the app.

Reply:
Can you leave a comment in these spots to that effect? Catch alls like this tend to stick once added because no one has the courage to remove. It helps if there's a breadcrumb.